staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go

```diff
@@ -351,6 +352,149 @@ func TestKMSv1Deprecation(t *testing.T) {
+				t.Fatalf("expected error %q, got %q", testCase.expectedErr, errString(err))
+			}
+
+		})
+	}
+	tts := []struct {
+		name            string
+		kmsv2Enabled    bool
+		expectedErr     string
+		expectedTimeout time.Duration
+		config          apiserverconfig.EncryptionConfiguration
+	}{
+		{
+			name:         "with kmsv1 and kmsv2, KMSv2=false",
+			kmsv2Enabled: false,
+			config: apiserverconfig.EncryptionConfiguration{
+				Resources: []apiserverconfig.ResourceConfiguration{
+					{
+						Resources: []string{"secrets"},
+						Providers: []apiserverconfig.ProviderConfiguration{
+							{
+								KMS: &apiserverconfig.KMSConfiguration{
+									Name:       "kms",
+									APIVersion: "v1",
+									Timeout: &metav1.Duration{
+										Duration: 1 * time.Second,
+									},
+									Endpoint:  "unix:///tmp/testprovider.sock",
+									CacheSize: pointer.Int32(1000),
+								},
+							},
+							{
+								KMS: &apiserverconfig.KMSConfiguration{
+									Name:       "another-kms",
+									APIVersion: "v2",
+									Timeout: &metav1.Duration{
+										Duration: 1 * time.Second,
+									},
+									Endpoint:  "unix:///tmp/anothertestprovider.sock",
+									CacheSize: pointer.Int32(1000),
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedErr: "KMSv2 feature is not enabled",
+		},
+		{
+			name:         "with kmsv1 and kmsv2, KMSv2=true",
+			kmsv2Enabled: true,
+			config: apiserverconfig.EncryptionConfiguration{
+				Resources: []apiserverconfig.ResourceConfiguration{
+					{
+						Resources: []string{"secrets"},
+						Providers: []apiserverconfig.ProviderConfiguration{
+							{
+								KMS: &apiserverconfig.KMSConfiguration{
+									Name:       "kms",
+									APIVersion: "v1",
+									Timeout: &metav1.Duration{
+										Duration: 1 * time.Second,
+									},
+									Endpoint:  "unix:///tmp/testprovider.sock",
+									CacheSize: pointer.Int32(1000),
+								},
+							},
+							{
+								KMS: &apiserverconfig.KMSConfiguration{
+									Name:       "another-kms",
+									APIVersion: "v2",
+									Timeout: &metav1.Duration{
+										Duration: 1 * time.Second,
+									},
+									Endpoint:  "unix:///tmp/anothertestprovider.sock",
+									CacheSize: pointer.Int32(1000),
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedErr: "",
+		},
+	}
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			// Just testing KMSv2 feature flag
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv1, true)()
+
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2, tt.kmsv2Enabled)()
+
+			ctx, cancel := context.WithCancel(context.Background())
+			cancel() // cancel this upfront so the kms v2 checks do not block
+
+			_, _, kmsUsed, err := getTransformerOverridesAndKMSPluginHealthzCheckers(ctx, &tt.config)
+			if err == nil {
+				assert.NotNil(t, kmsUsed)
+			}
```

Comment:
https://github.com/kubernetes/kubernetes/pull/119714#discussion_r1289079193 was not addressed, I would expect something like:

```suggestion
			_, _, kmsUsed, err := getTransformerOverridesAndKMSPluginHealthzCheckers(ctx, &tt.config)
			if err == nil {
				if kmsUsed == nil || kmsUsed.v2Used != wantV2Used { fail the test }
			}
```